companion/src/firmwares/moduledata.cpp

```diff
@@ -262,3 +268,274 @@ int ModuleData::getMaxChannelCount()
   }
   return 8;
 }
+
+/*
+      If Companion did not combine the /radio/src/datastructs type and sub-type fields
+      this would be so much cleaner and, modules and protocols generally easier to maintain
+      Another case for refactoring ModuleData as a whole!
+*/
+
+//  static
+int ModuleData::getTypeFromProtocol(unsigned int protocol)
+{
+  const QVector<QPair<int, int>>ProtocolTypeTable = {
+
+                          { PULSES_OFF, MODULE_TYPE_NONE },
+                          { PULSES_PPM, MODULE_TYPE_PPM },
+
+                          { PULSES_PXX_XJT_X16, MODULE_TYPE_XJT_PXX1 },
+                          { PULSES_PXX_XJT_D8, MODULE_TYPE_XJT_PXX1 },
+                          { PULSES_PXX_XJT_LR12, MODULE_TYPE_XJT_PXX1 },
+
+                          { PULSES_ACCESS_ISRM, MODULE_TYPE_ISRM_PXX2 },
+                          { PULSES_ACCST_ISRM_D16, MODULE_TYPE_ISRM_PXX2 },
+
+                          { PULSES_LP45, MODULE_TYPE_DSM2 },
+                          { PULSES_DSM2, MODULE_TYPE_DSM2 },
+                          { PULSES_DSMX, MODULE_TYPE_DSM2 },
+
+                          { PULSES_CROSSFIRE, MODULE_TYPE_CROSSFIRE },
+                          { PULSES_MULTIMODULE, MODULE_TYPE_MULTIMODULE },
+                          { PULSES_PXX_R9M, MODULE_TYPE_R9M_PXX1 },
+                          { PULSES_ACCESS_R9M, MODULE_TYPE_R9M_PXX2 },
+                          { PULSES_PXX_R9M_LITE, MODULE_TYPE_R9M_LITE_PXX1 },
+                          { PULSES_ACCESS_R9M_LITE, MODULE_TYPE_R9M_LITE_PXX2 },
+                          { PULSES_GHOST, MODULE_TYPE_GHOST },
+                          { PULSES_ACCESS_R9M_LITE_PRO, MODULE_TYPE_R9M_LITE_PRO_PXX2 },
+                          { PULSES_SBUS, MODULE_TYPE_SBUS },
+
+                          { PULSES_XJT_LITE_X16, MODULE_TYPE_XJT_LITE_PXX2 },
+                          { PULSES_XJT_LITE_D8, MODULE_TYPE_XJT_LITE_PXX2 },
+                          { PULSES_XJT_LITE_LR12, MODULE_TYPE_XJT_LITE_PXX2 },
+
+                          { PULSES_AFHDS3, MODULE_TYPE_FLYSKY }
+                      };
+
+  QPair<int, int>elmt;
+
+  foreach (elmt, ProtocolTypeTable) {
+    if (elmt.first == (int)protocol)
+      return elmt.second;
+  }
+
+  return -1;
+}
+
+//  static
+int ModuleData::getSubTypeFromProtocol(unsigned int protocol)
+{
+  if (protocol == PULSES_ACCST_ISRM_D16 || protocol == PULSES_ACCESS_ISRM)
+    return protocol - PULSES_ACCESS_ISRM;
+  else if (protocol >= PULSES_PXX_XJT_X16 && protocol <= PULSES_PXX_XJT_LR12)
+    return protocol - PULSES_PXX_XJT_X16;
+  else
+    return 0;
+}
+
+//  static
+QString ModuleData::typeToString(int type)
+{
+  static const char * strings[] = {
+    "OFF",
+    "PPM",
+    "FrSky XJT",
+    "ISRM",
+    "DSM2",
+    "TBS Crossfire",
```

Comment:
CRSF is not just TBS. ELRS uses CRSF as well, which could be confusing for users.